### PR TITLE
New CVar: `mp_dying_time`

### LINKED ISSUE #484: [feature request] Change time after death to switch camera
Need a new CVar
https://www.youtube.com/watch?v=9DAISSlRfz4

https://github.com/s1lentq/ReGameDLL_CS/blob/7983a9d9293b65d359b7971305362175ba3f277e/regamedll/dlls/player.cpp#L3602
m_fDeadTime + **3**

### PR DESCRIPTION
**Pull request by wopox1337**
close #484
```cpp
// Time for switch to free observing after death.
// 0 - disabled (instant view switch)
// >0.00001 - time delay to start spectate
//
// Default value: "3.0"
mp_dying_time 3.0
```
Video description: https://youtu.be/9DAISSlRfz4

### CODE REVIEW

File: regamedll/dlls/player.cpp [hunk 1]
```diff
@@ -2049,7 +2049,14 @@ void EXT_FUNC CBasePlayer::__API_HOOK(Killed)(entvars_t *pevAttacker, int iGib)
 	}
 
 	pev->modelindex = m_modelIndexPlayer;
+
+#ifdef REGAMEDLL_ADD
+	constexpr float DYING_MIN_TIME = 2.0f;
+	pev->deadflag = (CGameRules::GetDyingTime() <= DYING_MIN_TIME) ? DEAD_DEAD : DEAD_DYING;
```
Comment (afwn90cj93201nixr2e1re) [inline regamedll/dlls/player.cpp]: ?

Comment (wopox1337) [reply regamedll/dlls/player.cpp]: because the animation of death does not have time to play
Link with timecode: https://youtu.be/dX-Q7t0rFBk?t=36 (0:36)

**Comment by RauliTop** [comment]
mp_diecam_switch

It should be a better name

**Comment by wopox1337** [comment]
note: playtime depends on 🤔https://github.com/s1lentq/ReGameDLL_CS/blob/ab846f86f4ee19acbb31cc5da8d8108d2bb52dda/regamedll/dlls/player.cpp#L3735-L3739

**Comment by torriz** [comment]
Any reason this couldn't be merged?

**Comment by wopox1337** [comment]
@torriz need to solve some problems:
- https://youtu.be/dX-Q7t0rFBk?t=36

**Comment by RauliTop** [comment]
> @torriz need to solve some problems:
> 
> * https://youtu.be/dX-Q7t0rFBk?t=36

This isnt a problem. 

It can be used to killcam effect like in COD: https://callofduty.fandom.com/wiki/KillCam?file=Cod_camping.jpg

Let's go with this featureeee!!

**Comment by abecee** [comment]
Leggggggggggo. Can't wait to use this command...

**Comment by abecee** [comment]
The bug will be fixed ? When ? The player has 2 times animation for dying..

File: regamedll/dlls/gamerules.h [hunk 2]
```diff
@@ -46,6 +46,7 @@ const float ROUND_RESPAWN_TIME  = 20.0f;
 const float ROUND_BEGIN_DELAY   = 5.0f;	// delay before beginning new round
 const float ITEM_KILL_DELAY     = 300.0f;
 const float RADIO_TIMEOUT       = 1.5f;
+const float DYING_TIME          = 3.0f;
```
Comment (ShadowsAdi) [inline regamedll/dlls/gamerules.h:49]: `DYING_TIME` const could be renamed to `DEATH_ANIMATION_TIME`

**Comment by abecee** [comment]
@wopox1337 Can u reupload the files ?

**Comment by wopox1337** [comment]
Updated this branch to the current state. 

@abecee You can always download the updated ReGameDLL with these changes from the CI section (Github Actions project).
See artifacts for this branch: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/2604150677

**Comment by abecee** [comment]
> _Updated this branch with master branch changes._
> 
> @abecee You can always download the updated ReGameDLL with these changes from the CI section (Github Actions project). See artifacts for this branch: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/2604150677

I already have this in my pc, I'm just saying. Buuuuut is this gonna be in the next official verions of regame ? @wopox1337

**Comment by wopox1337** [comment]
@abecee This functionality is still incomplete. 

There is a problem with repeating the visibility of the player's death, as I showed earlier in the video: 
> @torriz need to solve some problems:
> 
> * https://youtu.be/dX-Q7t0rFBk?t=36

When this problem is solved, then it will be possible to add this functionality to the master branch.

**Comment by abecee** [comment]
> @abecee This functionality is still incomplete. 
> 
> There is a problem with repeating the visibility of the player's death, as I showed earlier in the video: 
> > @torriz need to solve some problems:
> > 
> > * https://youtu.be/dX-Q7t0rFBk?t=36
> 
> When this problem is solved, then it will be possible to add this functionality to the master branch.
> 
> 
> 

Oh, ok. This function is very very good and needed. Now I'm using a plugin for that but is not the same thing. When i was using the version with cam time and having that problem, my fix was with a plugin for hiding body's after they die and was good. @wopox1337

**Comment by abecee** [comment]
@wopox1337 u can try to do the same thing to hide the body after the first animation of dying and adding 2 cvars. 1 for enable/disable the function and another one for changing time 😁🤔

**Comment by dystopm** [comment]
> @abecee This functionality is still incomplete.
> 
> There is a problem with repeating the visibility of the player's death, as I showed earlier in the video:
> 
> > @torriz need to solve some problems:
> > 
> > * https://youtu.be/dX-Q7t0rFBk?t=36
> 
> When this problem is solved, then it will be possible to add this functionality to the master branch.

@wopox1337 I think I've found the solution for that bug, can you add me in the write access from this pull so I can commit the fix? I did a deep test and it was way easier than I thought

**Comment by wopox1337** [comment]
@dystopm You can [fork my repo and contribute a fix](https://github.com/wopox1337/ReGameDLL_CS/tree/death_cam_time) by creating an extra branch with this feature, and suggest in my PR repository.

**Comment by dystopm** [comment]
> @dystopm You can [fork my repo and contribute a fix](https://github.com/wopox1337/ReGameDLL_CS/tree/death_cam_time) by creating an extra branch with this feature, and suggest in my PR repository. Merge conflict now fixed

@wopox1337 done. 

I'll explain what I did: by just looking at the video, I immediately targeted ClCorpse message so I ended figuring out that issue was in the 8th argument which involves animtime and gametime, two common factors that triggered this bug. Value is always negative because `animtime` is set when `m_fSequenceFinished` is set to `true`, so in fact 3 seconds must pass and `animtime - gametime` will be -3, multiplied by 100, having a maximum value of -300. Bug ocurred when this value tended to zero, when animtime and gametime reached both a difference of 0 or 0.1. 

By just testing different values of mp_dying_time, and using HLMV to check model animations, I ended figuring out that `animtime - gametime` must be minimum the current sequence duration; everything else is explained in the code. I wasn't able to get an objective fix because I do not manage to have a decompiled CS client to see by myself how the 8th parameter of ClCorpse is handled. Anyway, get a drink and test it, I've already done it and it worked :)

**Comment by wopox1337** [comment]
@dystopm 

Decompiled part from CS
```cpp
int __cdecl CHudHealth::MsgFunc_ClCorpse(CHudHealth *const this, const char *pszName, int iSize, void *pbuf)
{
  const char *modelName; // ebx
  int v5; // esi
  int v6; // edi
  int v7; // ebp
  int v8; // eax
  long double v9; // fst7
  long double v10; // fst7
  char *format; // [esp+8h] [ebp-C4h]
  char *pModel; // [esp+Ch] [ebp-C0h]
  float pModela; // [esp+Ch] [ebp-C0h]
  float delay; // [esp+20h] [ebp-ACh]
  float v16; // [esp+20h] [ebp-ACh]
  float coordX; // [esp+34h] [ebp-98h]
  float vOrigin; // [esp+34h] [ebp-98h]
  float coordY; // [esp+38h] [ebp-94h]
  float vOrigin_4; // [esp+38h] [ebp-94h]
  float coordZ; // [esp+3Ch] [ebp-90h]
  float vOrigin_8; // [esp+3Ch] [ebp-90h]
  float vAngles; // [esp+40h] [ebp-8Ch]
  float vAngles_4; // [esp+44h] [ebp-88h]
  float vAngles_8; // [esp+48h] [ebp-84h]
  int v26; // [esp+4Ch] [ebp-80h]
  char szModel[64]; // [esp+58h] [ebp-74h] BYREF
  Vector p_vOrigin; // [esp+98h] [ebp-34h] BYREF
  Vector p_vAngles; // [esp+A4h] [ebp-28h] BYREF

  BEGIN_READ(pbuf, iSize);
  modelName = READ_STRING();
  coordX = (float)READ_LONG();
  vOrigin = 0.0078125 * coordX;
  coordY = (float)READ_LONG();
  vOrigin_4 = 0.0078125 * coordY;
  coordZ = (float)READ_LONG();
  vOrigin_8 = 0.0078125 * coordZ;
  vAngles = READ_COORD();
  vAngles_4 = READ_COORD();
  vAngles_8 = READ_COORD();
  v26 = READ_LONG();
  delay = ((long double (*)(void))gEngfuncs.GetClientTime)() + (long double)v26 / 100.0;
  v5 = READ_BYTE();
  v6 = READ_BYTE();
  v7 = READ_BYTE();
  v8 = READ_BYTE();
  v9 = delay;
  if ( !cl_minmodels || cl_minmodels->value == 0.0 )
  {
    if ( !strstr(modelName, "models/") )
    {
      snprintf(szModel, 0x40u, "models/player/%s/%s.mdl", modelName, modelName);
      v10 = delay;
      goto LABEL_10;
    }
    pModel = (char *)modelName;
    goto LABEL_9;
  }
  if ( v7 == 1 )
  {
    if ( !cl_min_t || (v9 = delay, !BIsValidTModelIndex((int)cl_min_t->value)) )
    {
      format = "models/player/leet/leet.mdl";
      goto LABEL_6;
    }
    pModel = sPlayerModelFiles[cl_min_t->value];
    goto LABEL_9;
  }
  if ( v7 == 2 )
  {
    if ( g_PlayerExtraInfo[v8].vip )
    {
      format = "models/player/vip/vip.mdl";
      goto LABEL_6;
    }
    if ( !cl_min_ct || (v9 = delay, !BIsValidCTModelIndex((int)cl_min_ct->value)) )
    {
      format = "models/player/gign/gign.mdl";
      goto LABEL_6;
    }
    pModel = sPlayerModelFiles[cl_min_ct->value];
LABEL_9:
    snprintf(szModel, 0x40u, "%s", pModel);
    v10 = delay;
    goto LABEL_10;
  }
  format = "models/player/player.mdl";
LABEL_6:
  v16 = v9;
  snprintf(szModel, 0x40u, format);
  v10 = v16;
LABEL_10:
  szModel[63] = 0;
  p_vAngles.x = vAngles;
  p_vAngles.y = vAngles_4;
  p_vAngles.z = vAngles_8;
  p_vOrigin.x = vOrigin;
  p_vOrigin.y = vOrigin_4;
  p_vOrigin.z = vOrigin_8;
  pModela = v10;
  CreateCorpse(&p_vOrigin, &p_vAngles, szModel, pModela, v5, v6);
  return 1;
}
```

**Comment by dystopm** [comment]
> ```c++
> delay = ((long double (*)(void))gEngfuncs.GetClientTime)() + (long double)v26 / 100.0;
> ```

Indeed here's the deal. 8th long value is negative, so that ended up guessing why anim started late again, if v26 (8th long, animtime - gametime) was an small value like 1, client-side corpse animtime will start too late.

Also, by just looking:

https://github.com/s1lentq/ReGameDLL_CS/blob/e1d1c11992e6f97037f6e6ab5baba015d2922db0/regamedll/dlls/player.cpp#L8842

That was a shitty way to put a higher value than current animation time; it should be directly the time it takes animation to end, so it will start at the last frame on client.

nevermind lmao

**Comment by wopox1337** [comment]
Still don't understand the final solution :D

**Comment by dystopm** [comment]
> Still don't understand the final solution :D

[Check your fork](https://github.com/wopox1337/ReGameDLL_CS/pull/12)

**Comment by wopox1337** [comment]
Nice one @dystopm! It remains to be tested well :)

**Comment by dystopm** [comment]
> Nice one @dystopm! It remains to be tested well :)
> 
> upd: tested now, without problems. https://youtube.com/live/e5Xlrjamq1M?feature=share

You're welcome. Just consider checking the next bunch of PRs I'll be adding :P

**Comment by etojuice** [comment]
> @wopox1337 I think I've found the solution for that bug, can you add me in the write access from this pull so I can commit the fix? I did a deep test and it was way easier than I thought

good catch!

**Comment by abecee** [comment]
Sooooooo it's finally happening? 🥹

**Comment by wopox1337** [comment]
> Sooooooo it's finally happening? 🥹

YES! We can ceremoniously pour drinks into glasses and celebrate the occasion through ingenuity @dystopm

**Comment by maxpain** [comment]
🎉🎉🎉

**Comment by abecee** [comment]
So let's see the UPDATE !!!!!!!!!

**Comment by dystopm** [comment]
> So let's see the UPDATE !!!!!!!!!

Wait until there's a considerable number of commits to make a new release. Instead, you can try the commit actions: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/5534763146 - select **linux32**

**Comment by Nord1cWarr1or** [comment]
Nice!

**Comment by abecee** [comment]
> > So let's see the UPDATE !!!!!!!!!
> 
> Wait until there's a considerable number of commits to make a new release. Instead, you can try the commit actions: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/5534763146 - select **linux32**

There are 2 Pre-releases versions so it's not an official version yet.

**Comment by wopox1337** [comment]
> > So let's see the UPDATE !!!!!!!!!
> 
> Wait until there's a considerable number of commits...

At least we should now finalize the use of `mp_fadetoblack` for this CVar

https://github.com/s1lentq/ReGameDLL_CS/pull/471#pullrequestreview-336502578

**Comment by abecee** [comment]
> > > So let's see the UPDATE !!!!!!!!!
> > 
> > 
> > Wait until there's a considerable number of commits...
> 
> At least we should now finalize the use of `mp_fadetoblack` for this CVar
> 
> [#471 (review)](https://github.com/s1lentq/ReGameDLL_CS/pull/471#pullrequestreview-336502578)

Fadetoblack is ...

**Comment by dystopm** [comment]
> > > So let's see the UPDATE !!!!!!!!!
> > 
> > 
> > Wait until there's a considerable number of commits to make a new release. Instead, you can try the commit actions: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/5534763146 - select **linux32**
> 
> There are 2 Pre-releases versions so it's not an official version yet.

Yeah but you can help us using this commit release and test it, so we can be sure of its implementation

**Comment by abecee** [comment]
> > > > So let's see the UPDATE !!!!!!!!!
> > > 
> > > 
> > > Wait until there's a considerable number of commits to make a new release. Instead, you can try the commit actions: https://github.com/s1lentq/ReGameDLL_CS/actions/runs/5534763146 - select **linux32**
> > 
> > 
> > There are 2 Pre-releases versions so it's not an official version yet.
> 
> Yeah but you can help us using this commit release and test it, so we can be sure of its implementation

Tested and works fine. No bugs.

**Comment by abecee** [comment]
Hi. I know I said there was no bug, but apparently there is one if you could call it that. It seems to take a little longer sometimes. For example I have the command set to 0, which should be "instant", but sometimes it's not, it seems like there's an extra second.
Any thoughts?

**Comment by dystopm** [comment]
> Hi. I know I said there was no bug, but apparently there is one if you could call it that. It seems to take a little longer sometimes. For example I have the command set to 0, which should be "instant", but sometimes it's not, it seems like there's an extra second. Any thoughts?

`0` - disable spectating around death.<br/>`>0.00001` - time delay to start spectate.<br/>**`NOTE`: The countdown starts when the player’s death animation is finished.**

**Comment by abecee** [comment]
> > Hi. I know I said there was no bug, but apparently there is one if you could call it that. It seems to take a little longer sometimes. For example I have the command set to 0, which should be "instant", but sometimes it's not, it seems like there's an extra second. Any thoughts?
> 
> `0` - disable spectating around death.<br/>`>0.00001` - time delay to start spectate.<br/>**`NOTE`: The countdown starts when the player’s death animation is finished.**

Yes, but sometimes it seems like is an extra second like i'm usig the cvar with 1 not 0

**Comment by dystopm** [comment]
> > > Hi. I know I said there was no bug, but apparently there is one if you could call it that. It seems to take a little longer sometimes. For example I have the command set to 0, which should be "instant", but sometimes it's not, it seems like there's an extra second. Any thoughts?
> > 
> > 
> > `0` - disable spectating around death.`>0.00001` - time delay to start spectate.**`NOTE`: The countdown starts when the player’s death animation is finished.**
> 
> Yes, but sometimes it seems like is an extra second like i'm usig the cvar with 1 not 0

That is because some death animations takes between 1.5 to 3.0 seconds to finish. Use HLMV and grab a player model and see their duration. If you are using custom animations they can take longer

**Comment by abecee** [comment]
> > > > Hi. I know I said there was no bug, but apparently there is one if you could call it that. It seems to take a little longer sometimes. For example I have the command set to 0, which should be "instant", but sometimes it's not, it seems like there's an extra second. Any thoughts?
> > > 
> > > 
> > > `0` - disable spectating around death.`>0.00001` - time delay to start spectate.**`NOTE`: The countdown starts when the player’s death animation is finished.**
> > 
> > Yes, but sometimes it seems like is an extra second like i'm usig the cvar with 1 not 0
> 
> That is because some death animations takes between 1.5 to 3.0 seconds to finish. Use HLMV and grab a player model and see their duration. If you are using custom animations they can take longer

Oh, ok. I was just asking. Ty for answer.

**Pull request merged**